Mathlib/RingTheory/Coalgebra/TensorProduct.lean

```diff
@@ -1,90 +1,160 @@
-[FLT repo](https://github.com/ImperialCollegeLondon/FLT/blob/eef74b4538c8852363936dfaad23e6ffba72eca5/FLT/mathlibExperiments/Coalgebra/TensorProduct.lean)
-is already universe polymorphic since it does not go via category theory.
+variable {R A B : Type*} [CommSemiring R] [AddCommMonoid B] [AddCommMonoid A]
+    [Module R A] [Module R B] [Coalgebra R A] [Coalgebra R B]
 
--/
+namespace TensorProduct
 
-universe v u
+open Coalgebra
 
-open CategoryTheory
-open scoped TensorProduct
+noncomputable
+instance instCoalgebraStruct : CoalgebraStruct R (A ⊗[R] B) where
+  comul :=
+    AlgebraTensorModule.tensorTensorTensorComm R R A A B B ∘ₗ
+      AlgebraTensorModule.map comul comul
+  counit := AlgebraTensorModule.rid R R R ∘ₗ AlgebraTensorModule.map counit counit
 
-section
+lemma comul_def :
+    Coalgebra.comul (R := R) (A := A ⊗[R] B) =
+      AlgebraTensorModule.tensorTensorTensorComm R R A A B B ∘ₗ
+        AlgebraTensorModule.map Coalgebra.comul Coalgebra.comul :=
+  rfl
 
-variable {R M N P Q : Type u} [CommRing R]
-  [AddCommGroup M] [AddCommGroup N] [Module R M] [Module R N] [Coalgebra R M] [Coalgebra R N]
-
-open MonoidalCategory in
-noncomputable instance TensorProduct.instCoalgebra : Coalgebra R (M ⊗[R] N) :=
-  let I := Monoidal.transport ((CoalgCat.comonEquivalence R).symm)
-  CoalgEquiv.toCoalgebra
-    (A := (CoalgCat.of R M ⊗ CoalgCat.of R N : CoalgCat R))
-    { LinearEquiv.refl R _ with
-      counit_comp := rfl
-      map_comp_comul := by
-        rw [CoalgCat.ofComonObjCoalgebraStruct_comul]
-        simp [-Mon_.monMonoidalStruct_tensorObj_X,
-          ModuleCat.MonoidalCategory.instMonoidalCategoryStruct_tensorHom,
-          ModuleCat.hom_comp, ModuleCat.of, ModuleCat.ofHom, comul_def,
-          ModuleCat.MonoidalCategory.tensorμ_eq_tensorTensorTensorComm] }
+@[deprecated (since := "2025-04-09")] alias instCoalgebraStruct_comul := comul_def
 
-end
+lemma counit_def :
+    Coalgebra.counit (R := R) (A := A ⊗[R] B) =
+      AlgebraTensorModule.rid R R R ∘ₗ AlgebraTensorModule.map counit counit :=
+  rfl
+
+@[deprecated (since := "2025-04-09")] alias instCoalgebraStruct_counit := counit_def
+
+@[simp]
+lemma comul_tmul (x : A) (y : B) : comul (x ⊗ₜ y) =
+  AlgebraTensorModule.tensorTensorTensorComm R R A A B B (comul x ⊗ₜ comul y) := rfl
+
+@[simp]
+lemma counit_tmul (x : A) (y : B) :
+  counit (R := R) (x ⊗ₜ[R] y) = counit (R := R) y • counit (R := R) x := rfl
+
+/-- `expand_comul R x with x₁ x₂` attempts to replace `comul (R := R) x` by
+`x₁ ⊗ₜ x₂` via linearity. -/
+scoped macro "expand_comul" ring:ident var:ident "with" var₁:ident var₂:ident : tactic =>
+  `(tactic|
+    (induction comul (R := $ring) $var with
+      | zero => simp only [tmul_zero, LinearEquiv.map_zero, LinearMap.map_zero,
+          zero_tmul, zero_mul, mul_zero]
+      | add _ _ h₁ h₂ => simp only [LinearEquiv.map_add, LinearMap.map_add,
+          tmul_add, add_tmul, add_mul, mul_add, h₁, h₂]
+      | tmul $var₁ $var₂ => ?_))
```

Comment:
I don't like this, but I have nothing better to suggest. It lies at a weird middle point between generality and specificity, so  I am happy to leave it like that

Reply:
#25197 has one version of that induction principle, but maybe it's more convenient to sum over a sigma type on Fin n, or an arbitrary fintype, instead.

Reply:
I sill want it to do `simp [map_sum, tmul_sum]` etc and then `sum_congr` and then `intro` so such macros will still be necessary.

Reply:
Right, you need
```lean
induction comul (R := R) x using TensorProduct.fin_sum_induction with | _ x₁ x₂;
      simp only [map_sum, sum_tmul, tmul_sum];
      congr 1;
      ext i : 1
```
and then you have `x₁ i` around in the goal instead of just `x₁`.

Reply:
I can see why such induction principles are useful but I am not convinced that they are better here. 
Another issue is my version produces `goals unsolved` when fails but this `congr` might make the goal unreadable when it doesn't do what it is intended to do.

Reply:
Yes; to be clear, I was agreeing that my solution doesn't seem to be be better.